// igniteui-angular/esm2022/lib/calendar/calendar.services.mjs
import { Injectable } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
export class KeyboardNavigationService {
    constructor(eventManager, ngZone) {
        this.eventManager = eventManager;
        this.ngZone = ngZone;
        this.keyHandlers = new Map();
        this.eventUnsubscribeFn = null;
    }
    attachKeyboardHandlers(elementRef, context) {
        this.detachKeyboardHandlers(); // Clean up any existing listeners
        this.ngZone.runOutsideAngular(() => {
            this.eventUnsubscribeFn = this.eventManager.addEventListener(elementRef.nativeElement, 'keydown', (event) => {
                const handler = this.keyHandlers.get(event.key);
                if (handler) {
                    this.ngZone.run(handler.bind(context, event));
                }
            });
        });
        return this;
    }
    detachKeyboardHandlers() {
        if (this.eventUnsubscribeFn) {
            this.eventUnsubscribeFn();
            this.eventUnsubscribeFn = null;
        }
        this.keyHandlers.clear();
    }
    set(key, handler) {
        this.keyHandlers.set(key, handler);
        return this;
    }
    unset(key) {
        this.keyHandlers.delete(key);
        return this;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: KeyboardNavigationService, deps: [{ token: i1.EventManager }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: KeyboardNavigationService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: KeyboardNavigationService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: i1.EventManager }, { type: i0.NgZone }] });
