// igniteui-angular/esm2022/lib/services/theme/theme.token.mjs
import { inject, InjectionToken } from "@angular/core";
import { mkenum } from "../../core/utils";
import { BehaviorSubject } from "rxjs";
import { DOCUMENT } from "@angular/common";
export class ThemeToken {
    constructor(t) {
        this.t = t;
        this.document = inject(DOCUMENT);
        const globalTheme = globalThis.window
            ?.getComputedStyle(this.document.body)
            .getPropertyValue("--ig-theme")
            .trim() || 'material';
        const _theme = t ?? globalTheme;
        this.subject = new BehaviorSubject(_theme);
    }
    onChange(callback) {
        return this.subject.subscribe(callback);
    }
    set(theme) {
        this.subject.next(theme);
    }
    get theme() {
        return this.subject.getValue();
    }
    get preferToken() {
        return !!this.t;
    }
}
export const THEME_TOKEN = new InjectionToken('ThemeToken', {
    providedIn: 'root',
    factory: () => new ThemeToken()
});
const Theme = /*@__PURE__*/ mkenum({
    Material: "material",
    Fluent: "fluent",
    Bootstrap: "bootstrap",
    IndigoDesign: "indigo",
});
