// igniteui-angular/esm2022/lib/combo/combo-add-item.component.mjs
import { IgxComboItemComponent } from './combo-item.component';
import { Component, HostBinding } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @hidden
 */
export class IgxComboAddItemComponent extends IgxComboItemComponent {
    get isDropDownItem() {
        return false;
    }
    get selected() {
        return false;
    }
    set selected(value) {
    }
    clicked(event) {
        this.comboAPI.disableTransitions = false;
        this.comboAPI.add_custom_item();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxComboAddItemComponent, deps: null, target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.4", type: IgxComboAddItemComponent, isStandalone: true, selector: "igx-combo-add-item", host: { properties: { "class.igx-drop-down__item": "this.isDropDownItem" } }, providers: [{ provide: IgxComboItemComponent, useExisting: IgxComboAddItemComponent }], usesInheritance: true, ngImport: i0, template: '<ng-content></ng-content>', isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxComboAddItemComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'igx-combo-add-item',
                    template: '<ng-content></ng-content>',
                    providers: [{ provide: IgxComboItemComponent, useExisting: IgxComboAddItemComponent }],
                    standalone: true
                }]
        }], propDecorators: { isDropDownItem: [{
                type: HostBinding,
                args: ['class.igx-drop-down__item']
            }] } });
