// igniteui-angular/esm2022/lib/directives/button/button-base.mjs
import { Directive, EventEmitter, HostBinding, HostListener, Input, Output, booleanAttribute } from '@angular/core';
import { mkenum } from '../../core/utils';
import * as i0 from "@angular/core";
export const IgxBaseButtonType = /*@__PURE__*/ mkenum({
    Flat: 'flat',
    Contained: 'contained',
    Outlined: 'outlined'
});
export class IgxButtonBaseDirective {
    /**
     * @hidden
     * @internal
     */
    onClick(ev) {
        this.buttonClick.emit(ev);
        this.focused = false;
    }
    /**
     * @hidden
     * @internal
     */
    onBlur() {
        this.focused = false;
    }
    /**
     * @hidden
     * @internal
     */
    get disabledAttribute() {
        return this.disabled || null;
    }
    constructor(element) {
        this.element = element;
        /**
         * Emitted when the button is clicked.
         */
        this.buttonClick = new EventEmitter();
        /**
         * Sets/gets the `role` attribute.
         *
         * @example
         * ```typescript
         * this.button.role = 'navbutton';
         * let buttonRole = this.button.role;
         * ```
         */
        this.role = 'button';
        /**
         * Sets/gets whether the button component is on focus.
         * Default value is `false`.
         * ```typescript
         * this.button.focus = true;
         * ```
         * ```typescript
         * let isFocused =  this.button.focused;
         * ```
         */
        this.focused = false;
        /**
          * Enables/disables the button.
          *
          * @example
          * ```html
          * <button igxButton="fab" disabled></button>
          * ```
          */
        this.disabled = false;
    }
    /**
     * @hidden
     * @internal
     */
    updateOnKeyUp(event) {
        if (event.key === "Tab") {
            this.focused = true;
        }
    }
    /**
     * Returns the underlying DOM element.
     */
    get nativeElement() {
        return this.element.nativeElement;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxButtonBaseDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "18.2.4", type: IgxButtonBaseDirective, inputs: { disabled: ["disabled", "disabled", booleanAttribute] }, outputs: { buttonClick: "buttonClick" }, host: { listeners: { "click": "onClick($event)", "blur": "onBlur()", "keyup": "updateOnKeyUp($event)" }, properties: { "attr.role": "this.role", "class.igx-button--focused": "this.focused", "class.igx-button--disabled": "this.disabled", "attr.disabled": "this.disabledAttribute" } }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxButtonBaseDirective, decorators: [{
            type: Directive
        }], ctorParameters: () => [{ type: i0.ElementRef }], propDecorators: { buttonClick: [{
                type: Output
            }], role: [{
                type: HostBinding,
                args: ['attr.role']
            }], onClick: [{
                type: HostListener,
                args: ['click', ['$event']]
            }], onBlur: [{
                type: HostListener,
                args: ['blur']
            }], focused: [{
                type: HostBinding,
                args: ['class.igx-button--focused']
            }], disabled: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }, {
                type: HostBinding,
                args: ['class.igx-button--disabled']
            }], disabledAttribute: [{
                type: HostBinding,
                args: ['attr.disabled']
            }], updateOnKeyUp: [{
                type: HostListener,
                args: ['keyup', ['$event']]
            }] } });
