// igniteui-angular/esm2022/lib/navigation-drawer/navigation-drawer.directives.mjs
import { Directive, HostBinding, Input, booleanAttribute } from '@angular/core';
import * as i0 from "@angular/core";
export class IgxNavDrawerItemDirective {
    constructor() {
        /**
         * Styles a navigation drawer item as selected.
         * If not set, `active` will have default value `false`.
         *
         * @example
         * ```html
         * <span igxDrawerItem [active]="true">Active Item</span>
         * ```
         */
        this.active = false;
        /**
         * Disables a navigation drawer item.
         * If not set, `disabled` will have default value `false`.
         *
         * @example
         * ```html
         * <span igxDrawerItem [disabled]="true">Disabled Item</span>
         * ```
         */
        this.disabled = false;
        /**
         * Styles a navigation drawer item as a group header.
         * If not set, `isHeader` will have default value `false`.
         *
         * @example
         * ```html
         * <span igxDrawerItem [isHeader]="true">Header</span>
         * ```
         */
        this.isHeader = false;
        /**
         * @hidden
         */
        this.activeClass = 'igx-nav-drawer__item--active';
        /**
        * @hidden
        */
        this.disabledClass = 'igx-nav-drawer__item--disabled';
    }
    /**
     * @hidden
     */
    get defaultCSS() {
        return !this.active && !this.isHeader;
    }
    /**
     * @hidden
     */
    get currentCSS() {
        return this.active && !this.isHeader && !this.disabled;
    }
    /**
     * @hidden
     */
    get headerCSS() {
        return this.isHeader;
    }
    /**
     * @hidden
     */
    get disabledCSS() {
        return this.disabled;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxNavDrawerItemDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "18.2.4", type: IgxNavDrawerItemDirective, isStandalone: true, selector: "[igxDrawerItem]", inputs: { active: ["active", "active", booleanAttribute], disabled: ["disabled", "disabled", booleanAttribute], isHeader: ["isHeader", "isHeader", booleanAttribute] }, host: { properties: { "class.igx-nav-drawer__item": "this.defaultCSS", "class.igx-nav-drawer__item--active": "this.currentCSS", "class.igx-nav-drawer__item--header": "this.headerCSS", "class.igx-nav-drawer__item--disabled": "this.disabledCSS" } }, exportAs: ["igxDrawerItem"], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxNavDrawerItemDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxDrawerItem]',
                    exportAs: 'igxDrawerItem',
                    standalone: true
                }]
        }], propDecorators: { active: [{
                type: Input,
                args: [{ alias: 'active', transform: booleanAttribute }]
            }], disabled: [{
                type: Input,
                args: [{ alias: 'disabled', transform: booleanAttribute }]
            }], isHeader: [{
                type: Input,
                args: [{ alias: 'isHeader', transform: booleanAttribute }]
            }], defaultCSS: [{
                type: HostBinding,
                args: ['class.igx-nav-drawer__item']
            }], currentCSS: [{
                type: HostBinding,
                args: ['class.igx-nav-drawer__item--active']
            }], headerCSS: [{
                type: HostBinding,
                args: ['class.igx-nav-drawer__item--header']
            }], disabledCSS: [{
                type: HostBinding,
                args: ['class.igx-nav-drawer__item--disabled']
            }] } });
export class IgxNavDrawerTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxNavDrawerTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.4", type: IgxNavDrawerTemplateDirective, isStandalone: true, selector: "[igxDrawer]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxNavDrawerTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxDrawer]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
export class IgxNavDrawerMiniTemplateDirective {
    constructor(template) {
        this.template = template;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxNavDrawerMiniTemplateDirective, deps: [{ token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.4", type: IgxNavDrawerMiniTemplateDirective, isStandalone: true, selector: "[igxDrawerMini]", ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxNavDrawerMiniTemplateDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[igxDrawerMini]',
                    standalone: true
                }]
        }], ctorParameters: () => [{ type: i0.TemplateRef }] });
